@john-bodley, I tried my best to add a test here, but wasn't able to get it to work due to an issue (maybe with FAB?). Here's my test code:
```python
    def test_annotation_json_endpoint(self):
        # Something is broken with annotations and RLS, teardown will undo this
        app.config["ENABLE_ROW_LEVEL_SECURITY"] = False

        self.login(username="admin")

        # Set up an annotation layer and annotation
        self.get_resp(
            "/annotationlayermodelview/add",
            {"form_data": json.dumps({"name": "foo", "descr": "bar"})},
        )
        self.get_resp(
            "/annotationmodelview/add",
            {
                "form_data": json.dumps(
                    {
                        "layer": "1",
                        "short_descr": "my_annotation",
                        "start_dttm": "2020-05-20 18:21:51",
                        "end_dttm": "2020-05-20 18:31:51",
                    }
                )
            },
        )

        resp = self.get_resp(
            "/superset/annotation_json/1?form_data=%7B%22time_range%22%3A%22100+years+ago+%3A+now%22%7D"
        )
        assert "my_annotation" in resp
```

And the error I saw:
```
======================================================================
ERROR: test_annotation_json_endpoint (tests.core_tests.CoreTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/erik_ritter/repos/github.com/incubator-superset/tests/core_tests.py", line 208, in test_annotation_json_endpoint
    "end_dttm": "2020-05-20 18:31:51",
  File "/Users/erik_ritter/repos/github.com/incubator-superset/tests/base_tests.py", line 191, in get_resp
    resp = self.client.post(url, data=data, follow_redirects=follow_redirects)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/werkzeug/test.py", line 1039, in post
    return self.open(*args, **kw)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/testing.py", line 227, in open
    follow_redirects=follow_redirects,
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/werkzeug/test.py", line 993, in open
    response = self.run_wsgi_app(environ.copy(), buffered=buffered)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/werkzeug/test.py", line 884, in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/werkzeug/test.py", line 1119, in run_wsgi_app
    app_rv = app(environ, start_response)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/security/decorators.py", line 109, in wraps
    return f(self, *args, **kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/views.py", line 589, in add
    self.add_template, title=self.add_title, widgets=widget
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/baseviews.py", line 281, in render_template
    template, **dict(list(kwargs.items()) + list(self.extra_args.items()))
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/templating.py", line 140, in render_template
    ctx.app,
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask/templating.py", line 120, in _render
    rv = template.render(context)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/general/model/add.html", line 2, in top-level template code
    {% import 'appbuilder/general/lib.html' as lib %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/base.html", line 1, in top-level template code
    {% extends base_template %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/superset/templates/superset/base.html", line 20, in top-level template code
    {% from 'superset/partials/asset_bundle.html' import css_bundle, js_bundle with context %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/superset/templates/appbuilder/baselayout.html", line 20, in top-level template code
    {% import 'appbuilder/baselib.html' as baselib %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/init.html", line 46, in top-level template code
    {% block body %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/superset/templates/appbuilder/baselayout.html", line 39, in block "body"
    {% block content %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/general/model/add.html", line 7, in block "content"
    {% block add_form %}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/general/model/add.html", line 8, in block "add_form"
    {{ widgets.get('add')(form_action=form_action)|safe }}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/widgets.py", line 37, in __call__
    return template.render(args)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/general/widgets/form.html", line 45, in top-level template code
    {{ lib.render_field(field, begin_sep_label, end_sep_label, begin_sep_field, end_sep_field) }}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/jinja2/runtime.py", line 574, in _invoke
    rv = self._func(*arguments)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/templates/appbuilder/general/lib.html", line 233, in template
    {{ field(**kwargs)|safe }}
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/wtforms/fields/core.py", line 155, in __call__
    return self.meta.render_field(self, kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/wtforms/meta.py", line 56, in render_field
    return field.widget(field, **render_kw)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/flask_appbuilder/fieldwidgets.py", line 156, in __call__
    return super(Select2Widget, self).__call__(field, **kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/wtforms/widgets/core.py", line 324, in __call__
    html.append(self.render_option(val, label, selected))
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/wtforms/widgets/core.py", line 337, in render_option
    return HTMLString('<option %s>%s</option>' % (html_params(**options), escape_html(label, quote=False)))
  File "/Users/erik_ritter/repos/github.com/incubator-superset/.tox/py36/lib/python3.6/site-packages/wtforms/widgets/core.py", line 31, in escape_html
    s = escape(text_type(s), quote=quote)
TypeError: __str__ returned non-string (type NoneType)
```

It looks like when I'm trying to create an annotation inside the annotation layer, it's rendering a template instead of creating the new annotation. Maybe @dpgaspar has thoughts here?

--- a/tests/core_tests.py
+++ b/tests/core_tests.py
@@ -25,6 +25,7 @@ import json
 import logging
 import os
 from typing import Dict, List, Optional
+from urllib.parse import quote
 
 import pytz
 import random
@@ -50,6 +51,7 @@ from superset.datasets.dao import DatasetDAO
 from superset.db_engine_specs.base import BaseEngineSpec
 from superset.db_engine_specs.mssql import MssqlEngineSpec
 from superset.models import core as models
+from superset.models.annotations import Annotation, AnnotationLayer
 from superset.models.dashboard import Dashboard
 from superset.models.datasource_access_request import DatasourceAccessRequest
 from superset.models.slice import Slice
@@ -76,12 +78,10 @@ class CoreTests(SupersetTestCase):
             tbl.table_name: tbl.id for tbl in (db.session.query(SqlaTable).all())
         }
         self.original_unsafe_db_setting = app.config["PREVENT_UNSAFE_DB_CONNECTIONS"]
-        self.original_rls_setting = app.config["ENABLE_ROW_LEVEL_SECURITY"]
 
     def tearDown(self):
         db.session.query(Query).delete()
         app.config["PREVENT_UNSAFE_DB_CONNECTIONS"] = self.original_unsafe_db_setting
-        app.config["ENABLE_ROW_LEVEL_SECURITY"] = self.original_rls_setting
 
     def test_login(self):
         resp = self.get_resp("/login/", data=dict(username="admin", password="general"))
@@ -180,33 +180,26 @@ class CoreTests(SupersetTestCase):
         assert '"Jennifer"' in resp
 
     def test_annotation_json_endpoint(self):
-        # Something is broken with annotations and RLS, teardown will undo this
-        app.config["ENABLE_ROW_LEVEL_SECURITY"] = False
-
-        self.login(username="admin")
-
         # Set up an annotation layer and annotation
-        self.get_resp(
-            "/annotationlayermodelview/add",
-            {"form_data": json.dumps({"name": "foo", "descr": "bar"})},
+        layer = AnnotationLayer(name="foo", descr="bar")
+        db.session.add(layer)
+        db.session.commit()
+
+        annotation = Annotation(
+            layer_id=layer.id,
+            short_descr="my_annotation",
+            start_dttm=datetime.datetime(2020, 5, 20, 18, 21, 51),
+            end_dttm=datetime.datetime(2020, 5, 20, 18, 31, 51),
         )
-        self.get_resp(
-            "/annotationmodelview/add",
-            {
-                "form_data": json.dumps(
-                    {
-                        "layer": "1",
-                        "short_descr": "my_annotation",
-                        "start_dttm": "2020-05-20 18:21:51",
-                        "end_dttm": "2020-05-20 18:31:51",
-                    }
-                )
-            },
-        )
+
+        db.session.add(annotation)
+        db.session.commit()
 
         resp = self.get_resp(
-            "/superset/annotation_json/1?form_data=%7B%22time_range%22%3A%22100+years+ago+%3A+now%22%7D"
+            f"/superset/annotation_json/{layer.id}?form_data="
+            + quote(json.dumps({"time_range": "100 years ago : now"}))
         )
+
         assert "my_annotation" in resp
 
     def test_old_slice_csv_endpoint(self):

--- a/tests/core_tests.py
+++ b/tests/core_tests.py
@@ -76,10 +76,12 @@ class CoreTests(SupersetTestCase):
             tbl.table_name: tbl.id for tbl in (db.session.query(SqlaTable).all())
         }
         self.original_unsafe_db_setting = app.config["PREVENT_UNSAFE_DB_CONNECTIONS"]
+        self.original_rls_setting = app.config["ENABLE_ROW_LEVEL_SECURITY"]
 
     def tearDown(self):
         db.session.query(Query).delete()
         app.config["PREVENT_UNSAFE_DB_CONNECTIONS"] = self.original_unsafe_db_setting
+        app.config["ENABLE_ROW_LEVEL_SECURITY"] = self.original_rls_setting
 
     def test_login(self):
         resp = self.get_resp("/login/", data=dict(username="admin", password="general"))
@@ -176,6 +178,36 @@ class CoreTests(SupersetTestCase):
         slc = self.get_slice("Girls", db.session)
         resp = self.get_resp(slc.explore_json_url)
         assert '"Jennifer"' in resp
+
+    def test_annotation_json_endpoint(self):
+        # Something is broken with annotations and RLS, teardown will undo this
+        app.config["ENABLE_ROW_LEVEL_SECURITY"] = False
+
+        self.login(username="admin")
+
+        # Set up an annotation layer and annotation
+        self.get_resp(
+            "/annotationlayermodelview/add",
+            {"form_data": json.dumps({"name": "foo", "descr": "bar"})},
+        )
+        self.get_resp(
+            "/annotationmodelview/add",
+            {
+                "form_data": json.dumps(
+                    {
+                        "layer": "1",
+                        "short_descr": "my_annotation",
+                        "start_dttm": "2020-05-20 18:21:51",
+                        "end_dttm": "2020-05-20 18:31:51",
+                    }
+                )
+            },
+        )
+
+        resp = self.get_resp(
+            "/superset/annotation_json/1?form_data=%7B%22time_range%22%3A%22100+years+ago+%3A+now%22%7D"
+        )
+        assert "my_annotation" in resp
 
     def test_old_slice_csv_endpoint(self):
         self.login(username="admin")

--- a/tests/superset_test_config.py
+++ b/tests/superset_test_config.py
@@ -76,3 +76,5 @@ CELERY_CONFIG = CeleryConfig
 CUSTOM_TEMPLATE_PROCESSORS = {
     CustomPrestoTemplateProcessor.engine: CustomPrestoTemplateProcessor
 }
+
+PRESERVE_CONTEXT_ON_EXCEPTION = False

